File: whitespace-java/src/main/java/dev/whitespace/service/bundler/Bundler.java
package dev.whitespace.service.bundler;

import dev.whitespace.service.bundler.domain.Recordable;
import dev.whitespace.service.bundler.resolver.ReferenceResolver;
import dev.whitespace.service.bundler.scanner.RecordableScanner;
import lombok.RequiredArgsConstructor;

import java.util.List;

@RequiredArgsConstructor
public class Bundler {
	private final List<Recordable> recordables;

	private final ReferenceResolver referenceResolver = new ReferenceResolver();

    void execute() {
		var refs = new RecordableScanner(recordables).find();
		var result = referenceResolver.resolve(refs);
		IO.println(result);
    }
}

File: whitespace/src/main/kotlin/dev/whitespace/service/bundler/Bundler.kt
package dev.whitespace.service.bundler

import dev.whitespace.domain.recording.Recordable
import io.quarkus.logging.Log
import java.nio.file.Path
import kotlinx.serialization.json.Json
import java.lang.Exception
import kotlin.io.path.readLines

class Bundler(private val input: Input) {
    suspend fun run() {
    }
}


data class Input(val recordables: List<Recordable>) {
    companion object {
        val json = Json {
            ignoreUnknownKeys = true
            explicitNulls = false
        }

        fun fromPath(path: Path): Input {
            val recordables = path.readLines(Charsets.UTF_8).mapIndexed { i, r ->
                try {
                    json.decodeFromString<Recordable>(r)
                } catch (e: Exception) {
                    Log.error("error when parsing line at index $i")
                    throw e
                }
            }

            return Input(recordables)
        }
    }
}

